doc(facade/PC): add README (#5123)
MovingBlocks/Terasology: facades/PC/README.md

profile - Run 'Terasology' to play the game as a standard PC application (with Java FlightRecorder profiling)
server - Starts a headless multiplayer server with data stored in [project-root]/terasology-server
```

To pass additional arguments to the PC Facade when running via Gradle, use the `--args` option (see the Gradle [Application Plugin](https://docs.gradle.org/6.9.1/userguide/application_plugin.html) documentation).
For instance, to print the usage help, run

```sh
gradlew game --args='--help'
```

## Architecture

The main entry point for the PC Facade is `Terasology#main`.
Starting from there, the game engine is initialized and started according to the user inputs.

<figure style="border:1px #cccccc solid; padding:4px;">
  <!-- this image is an editable SVG created with Draw.io -->
  <img 
    style="display:block; margin-left:auto; margin-right:auto"
    src="./docs/pc-facade-overview.drawio.svg" 
    alt="Architectural overview of the PC Facade entry point."/>
  <figcaption style="text-align:center;">Architectural overview of the PC Facade entry point.</figcaption>

> </figure>

Author: narg, fancy CSS does not work on Github :pensive: guess I'll use a good ol' `align="center"` somewhere then...

Author: ```suggestion
<figure align="center">
  <!-- this image is an editable SVG created with Draw.io -->
  <img 
    src="./docs/pc-facade-overview.drawio.svg" 
    alt="Architectural overview of the PC Facade entry point."/>
  <figcaption>Architectural overview of the PC Facade entry point.</figcaption>
</figure>
```

# PC Facade

> The _PC Facade_ is the front-facing layer for running _Terasology_ on the PC.

Reviewer: ```suggestion
The _PC Facade_ is the front-facing layer for running _Terasology_ on a PC.
```
(Although this change might be a personal preference.)
It might also be worth mentioning that "PC" (Personal Computer?) refers to the Windows, macOS and Linux platforms specifically.

Reviewer 2: @BenjaminAmos How about the following phrasing?
```suggestion
The _PC Facade_ is the front-facing layer for running _Terasology_ on a personal computer (PC) with a Windows, macOS, or Linux operating system.
```

Reviewer: I've changed my mind about expanding the personal computer abbreviation. I don't think any changes are necessary here anymore. New OSes might come along and we're not doing too much platform-specific stuff ourselves (I think). Terasology's supported platforms are well documented elsewhere.

```suggestion
The _PC Facade_ is the front-facing layer for running _Terasology_ on computers running Windows, macOS, or Linux.
```

## Usage

Most users won't interact directly with the PC Facade, but instead use the [Terasology Launcher](https://github.com/MovingBlocks/TerasologyLauncher) to start the game.

When starting distributions of Terasology manually, users will interact with the command-line interface ([CLI](#cli)).

Developers will most often use the PC Facade via [Gradle](#gradle) or via a run configuration in their IDE.

### CLI

The PC Facade provides a command-line interface (CLI) for configuring Terasology.

> See the usage help (`./Terasology --help`) for a full overview.

Reviewer: Is it worth mentioning that the CLI options are managed through the [`Terasology`](https://github.com/MovingBlocks/Terasology/blob/26d5b5237ba34b4b4c97d9f10469512432b14d90/facades/PC/src/main/java/org/terasology/engine/Terasology.java) class using [PicoCLI](https://picocli.info/)?

Reviewer 2: Yeah why not... How about the following:
```suggestion
See the usage help (`./Terasology --help`) for a full overview.
For implementing the CLI, we make use of the [PicoCLI](https://picocli.info/) framework for Java command-line applications.
```

Reviewer: That wording works fine. It's just a bit of useful context in-case somebody wanted to know how to add a new command line option, for example.

debug - Run 'Terasology' to play the game as a standard PC application (in debug mode)
game - Run 'Terasology' to play the game as a standard PC application
permissiveNatives - Run 'Terasology' with security set to permissive and natives loading a second way (for KComputers)
profile - Run 'Terasology' to play the game as a standard PC application (with Java FlightRecorder profiling)
server - Starts a headless multiplayer server with data stored in [project-root]/terasology-server
```

To pass additional arguments to the PC Facade when running via Gradle, use the `--args` option (see the Gradle [Application Plugin](https://docs.gradle.org/6.9.1/userguide/application_plugin.html) documentation).
For instance, to print the usage help, run

```sh
gradlew game --args='--help'
```

## Architecture

The main entry point for the PC Facade is `Terasology#main`.
Starting from there, the game engine is initialized and started according to the user inputs.

<p align="center">
<!-- this image is an editable SVG created with Draw.io -->
<img src="./docs/pc-facade-overview.drawio.svg" alt="Architectural overview of the PC Facade entry point."/>
</p>

The execution flow is as follows:

> 1. A `TerasologyEngineBuilder` is instantiated to preparet the game engine.

Reviewer: ```suggestion
1. A `TerasologyEngineBuilder` is instantiated to prepare the game engine.
```

> 1. Additional subsystems are collected via the builder. This creates and adds subsystems that are specific for running Terasology on a PC. The facade also chooses respective subsystem implementations depending on whether the game is run in _headless mode_.

Reviewer: > The facade also chooses respective subsystem implementations depending on whether the game is run in _headless mode_.

It might be worth providing an example of this, for example using `HeadlessGraphics` instead of `LwjglGraphics`.

Reviewer 2: @BenjaminAmos Would a simple
```suggestion
1. Additional subsystems are collected via the builder. This creates and adds subsystems that are specific for running Terasology on a PC. The facade also chooses respective subsystem implementations depending on whether the game is run in _headless mode_. For instance, for a headless server it would choose `HeadlessGraphics` instead of `LwjglGraphics`.
```
suffice or would you like a more detailed example?

Reviewer: That level of detail should be enough.

> 1. The builder is used to create an instance of a `TerasologyEngine` which is subsequently initalized (`TerasologyEngine#initialize`, not to be confused with `GameEngine#initializeRun`). This trigger initialization of the engine for subsystems and asset managemet. This does not start the game loop yet.

Reviewer: ```suggestion
1. The builder is used to create an instance of `TerasologyEngine` which is subsequently initialized (`TerasologyEngine#initialize`, not to be confused with `GameEngine#initializeRun`). This triggers initialization of the engine for subsystems and asset management. This does not start the game loop yet.
```

1. Depending on the arguments provided by the user, the facade creates an appropriate `GameState`
    - `HeadlessSetup` if the game is started in _headless mode_

>     - `Loading` if a game should diretly be loaded (either by loading a save game or by creating a new one)

Reviewer: ```suggestion
    - `Loading` if a game should directly be loaded (either by loading a save game or by creating a new one)
```
